proto/v1beta1/intoto.proto

```diff
@@ -0,0 +1,145 @@
+// Copyright 2020 The Grafeas Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package grafeas.v1beta1.intoto;
+
+option go_package = "github.com/grafeas/grafeas/proto/v1beta1/intoto_go_proto";
+option java_multiple_files = true;
+option java_package = "io.grafeas.v1beta1.intoto";
+option objc_class_prefix = "GRA";
+
+import "google/protobuf/struct.proto";
+
+// This file defines the proto that matches the schema of an in-toto link
+// file. in-toto is a framework to protect a software supply chain. You can read
+// more about in-toto at https://in-toto.io/ and you can find the current
+// version of the in-toto specification at
+// https://github.com/in-toto/docs/blob/master/in-toto-spec.md
+
+
+// A signature object consists of the KeyID used and the signature itself.
+message Signature {
+  string keyid = 1 [json_name="keyid"];
+  string sig = 2 [json_name="sig"];
+}
+
+// An object defined to capture the by products of a step.
+message ByProductObject {
+  int32 return_value = 1 [json_name="return-value"];
+  string stderr = 2 [json_name="stderr"];
+  string stdout = 3 [json_name="stdout"];
+}
+
+// This corrresponds to the core in-toto link attestation.
+message Link {
+  // This field contains the full command executed for the step. This can also
+  // be empty if attestations are generated for operations that aren't directly
+  // mapped to a specific command. Each term in the command is an independent
+  // string in the list. An example of a command in the in-toto metadata field
+  // is:
+  // "command": ["git", "clone", "https://github.com/in-toto/demo-project.git"]
+  repeated string command = 1 [json_name="command"];
+
+  // Materials are the supply chain artifacts that go into the step and are used
+  // for the operation performed. The key of the map is the path of the artifact
+  // and the structure contains the recorded hash information. An example is:
+  // "materials": {
+  //   "foo/bar": {
```

Comment:
Could we change this to one of the formats described [here](https://github.com/grafeas/grafeas/blob/master/docs/grafeas_concepts.md#resource-urls)?

Reply:
We could use the format specified there for the key of a materials/products object (however this will require ITE-4: https://github.com/in-toto/ITE/pull/6/files#diff-669c59f42864911b31bda53fe6b3d41e) but the hash object allows us to support non SHA-256 hashes as well as multiple hash algorithms for a particular artifact.

Reply:
I think it's wortwhile keeping the hash-objects, and then we can massage both implementations later down the line to avoid redundancy. I'm curious to know if there's a way to define hash values in such a way that you can leverage multiple-hashing algorithms at once. 

As an aside, some of these definitions (e.g., `pip`) may could be improved to include namespacing and resource-location. I don't think the current schema supports uses like self-hosted pypi registries and more. In the end, pip is just a frontend to the pypi/warehouse/other package repositories...

Reply:
I'm confused. What I'm wondering is what does `foo/bar` represent here? It sounds like it's supposed to represent an artifact, based on `The key of the map is the path of the artifact`. If so, we represent all artifacts using CPE URIs, and I'd like to keep this consistent. In Grafeas, the artifacts are binaries, packages, containers, etc. Do we have the same definition here?

```
"foo/bar": {
  "sha256": "ebebf..."
   }
}
```

Reply:
`foo/bar` represents a path in the file system to individual components that make up a software package as they move through the supply chain. This is going to be more generalised with the work in [ITE-4](https://github.com/in-toto/ITE/pull/6) which allows for URIs that refer to more abstract artifacts.

As I understand the CPE specification, I'm not seeing a lot of overlap between our artifact URIs and CPE URIs - they don't represent the components within the package's supply chain but rather the package as a whole. Am I misunderstanding the spec? Please let me know :)

Reply:
Ok, great, there *is* a way to represent files in Grafeas as well, please see [resource urls](https://github.com/grafeas/grafeas/blob/master/docs/grafeas_concepts.md#resource-urls), `Generic file`.

Yes, the idea here is that an artifact in Grafeas is a file, and a package in which the file is a component, and the container that might contain packages. Here's an example of tracking software supply chain (SSC) for an individual file:

file (eg source code, test, data etc) -> built into a package -> tested -> deployed.

So we can track the lifetime of a file through SSC if we need to, it can even be at finer granularity if we need it, eg block of code, etc. CPE specification is just a generic way to represent this. Does this make sense?

Reply:
I think the confusion is that in in-toto, the `foo/bar` points to a map with hash values for `foo/bar`. The `Generic File` URI contains very similar information except it includes everything in a single string. The difference is that the map allows us to record more than one hash, using more than one algorithm simultaneously.

Are you suggesting we update `foo/bar` to the `Generic File` URI format and have it point to the map with the hash values? Something like:

```
"file://sha256:244fd47e07d1004f0aed9c156aa09083c82bf8944eceb67c946ff7430510a77b:foo/bar" {
  "sha256": "244fd47e07d1004f0aed9c156aa09083c82bf8944eceb67c946ff7430510a77b",
}
```

Reply:
It could be `"file://sha256:244fd47e07d1004f0aed9c156aa09083c82bf8944eceb67c946ff7430510a77b:foo/bar"` and have it build in the hash inside, or `"file://foo/bar"` etc.